pallets/kensetsu/src/lib.rs

```diff
@@ -1149,49 +1140,76 @@ pub mod pallet {
-                        )
-                        .map_err(|_| Error::<T>::ArithmeticError)?;
-                        collateral_info.last_fee_update_time = now;
-                        collateral_info.interest_coefficient =
-                            FixedU128::from_inner(new_coefficient);
-                    }
-                    Ok::<CollateralInfo<T::Moment>, DispatchError>(collateral_info.clone())
-                })?;
+            let mut collateral_info = CollateralInfos::<T>::get(collateral_asset_id)
+                .ok_or(Error::<T>::CollateralInfoNotFound)?;
+            let now = Timestamp::<T>::get();
+            ensure!(
+                now >= collateral_info.last_fee_update_time,
+                Error::<T>::AccrueWrongTime
+            );
 
+            // do not update if time is the same
+            if now > collateral_info.last_fee_update_time {
+                let time_passed = now
+                    .checked_sub(&collateral_info.last_fee_update_time)
+                    .ok_or(Error::<T>::ArithmeticError)?;
+                let new_coefficient = compound(
+                    collateral_info.interest_coefficient.into_inner(),
+                    collateral_info.risk_parameters.stability_fee_rate,
+                    time_passed
+                        .checked_into::<u64>()
+                        .ok_or(Error::<T>::ArithmeticError)?,
+                )
+                .map_err(|_| Error::<T>::ArithmeticError)?;
+                collateral_info.last_fee_update_time = now;
+                collateral_info.interest_coefficient = FixedU128::from_inner(new_coefficient);
+            }
             Ok(collateral_info)
         }
 
+        /// Recalculates and updates collateral interest coefficient with the current timestamp
+        fn update_collateral_interest_coefficient(
+            collateral_asset_id: &AssetIdOf<T>,
+        ) -> Result<CollateralInfo<T::Moment>, DispatchError> {
+            let updated_collateral_info =
+                Self::calculate_collateral_interest_coefficient(collateral_asset_id)?;
+            <CollateralInfos<T>>::set(collateral_asset_id, Some(updated_collateral_info.clone()));
+            Ok(updated_collateral_info)
+        }
+
+        /// Calculates stability fee for the CDP for the current time.
+        fn calculate_stability_fee(cdp_id: CdpId) -> Result<Balance, DispatchError> {
+            let cdp = Self::cdp(cdp_id).ok_or(Error::<T>::CDPNotFound)?;
+            let collateral_info =
+                Self::calculate_collateral_interest_coefficient(&cdp.collateral_asset_id)?;
+            let interest_coefficient = collateral_info.interest_coefficient;
+            let interest_percent = interest_coefficient
+                .checked_sub(&cdp.interest_coefficient)
+                .ok_or(Error::<T>::ArithmeticError)?
+                .checked_div(&cdp.interest_coefficient)
+                .ok_or(Error::<T>::ArithmeticError)?;
```

Comment:
As I see, `interest_coefficient` is `FixedU128`.
Is it possible that `cdp.interest_coefficient > collateral_info.interest_coefficient` ?
If yes, it means the `checked_sub` returns error, is this intended?
If we don't need the error, probably we can use `saturating_sub` that returns 0 for this case.

Reply:
Also `checked_div` returns error for `cdp.interest_coefficient = 0`.
Do we want error or 0 as a result?

Reply:
`interest_coefficient` always initialized with `1` and growth over time by stability percent. `collateral_info` coefficient is recalculated and assigned to cdp.interest_coefficient. Thus, `collateral_info.interest_coefficient` >= `cdp.interest_coefficient` >= 1.
If there is something wrong, we want to get an error.

Reply:
Ok